AsyncDisplayKit/Private/_ASHierarchyChangeSet.m

```diff
@@ -198,9 +198,41 @@ - (void)_sortAndCoalesceChangeArrays
 
     // Ignore item inserts in reloaded(new)/inserted sections.
     [_ASHierarchyItemChange sortAndCoalesceChanges:_insertItemChanges ignoringChangesInSections:insertedOrReloaded];
+    
+
+    // reload itemsChanges need to be adjusted so that we access the correct indexPaths in the datasource
+    NSDictionary *insertedIndexPathsMap = [_ASHierarchyItemChange sectionToIndexSetMapFromChanges:_insertItemChanges ofType:_ASHierarchyChangeTypeInsert];
+    NSDictionary *deletedIndexPathsMap = [_ASHierarchyItemChange sectionToIndexSetMapFromChanges:_deleteItemChanges ofType:_ASHierarchyChangeTypeDelete];
+
+    for (_ASHierarchyItemChange *change in _reloadItemChanges) {
+      NSAssert(change.changeType == _ASHierarchyChangeTypeReload, @"It must be a reload change to be in here");
+      NSMutableArray *newIndexPaths = [NSMutableArray array];
+      
+      // Every indexPaths in the change need to update its section and/or row
+      // depending on all the deletions and insertions
+      for (NSIndexPath *indexPath in change.indexPaths) {
+        NSUInteger section = indexPath.section;
+        NSUInteger row = indexPath.row;
+        
+        section -= [_deletedSections countOfIndexesInRange:NSMakeRange(0, section)];
+        section += [_insertedSections countOfIndexesInRange:NSMakeRange(0, section)];
+        
+        NSIndexSet *indicesInsertedInSection = insertedIndexPathsMap[@(section).stringValue];
+        row += [indicesInsertedInSection countOfIndexesInRange:NSMakeRange(0, row)];
+        NSIndexSet *indicesDeletedInSection = deletedIndexPathsMap[@(section).stringValue];
+        row += [indicesDeletedInSection countOfIndexesInRange:NSMakeRange(0, row)];
```

Comment:
@Adlai-Holler absolutely! I realized these issues as I did more testing. addressing them now. I also found other issues with reloads as we just delete+insert them, and after the nodes are completed, we call the super methods on the UICollectionView. This means that the delete indexPaths should be current and insert indexPaths should be future, which will be different in many cases, but we are passing through the same set.

I will have a PR soon.